src/Avalonia.Controls/Documents/Span.cs

```diff
@@ -68,26 +70,26 @@ internal override void OnInlineHostChanged(IInlineHost? oldValue, IInlineHost? n
         {
             base.OnInlineHostChanged(oldValue, newValue);
 
-            if (Inlines is not null)
-            {
-                Inlines.InlineHost = newValue;
-            }
+            Inlines.InlineHost = newValue;
         }
 
         private void OnInlinesChanged(InlineCollection? oldValue, InlineCollection? newValue)
         {
+            void OnInlinesInvalidated(object? sender, EventArgs e)
+                => InlineHost?.Invalidate();
+
             if (oldValue is not null)
             {
                 oldValue.LogicalChildren = null;
                 oldValue.InlineHost = null;
-                oldValue.Invalidated -= (s, e) => InlineHost?.Invalidate();
+                oldValue.Invalidated -= OnInlinesInvalidated;
             }
 
             if (newValue is not null)
             {
                 newValue.LogicalChildren = LogicalChildren;
                 newValue.InlineHost = InlineHost;
-                newValue.Invalidated += (s, e) => InlineHost?.Invalidate();
+                newValue.Invalidated += OnInlinesInvalidated;
```

Comment:
Feels like it was a bug before.

Reply:
Definitely, the `-=` above didn't do anything.